core/controllers/suggestion.py

```diff
@@ -264,17 +264,86 @@ def post(self) -> None:
             )
             if new_image_filenames and files is not None:
-                _upload_suggestion_images(
+                self._upload_new_images_in_translation(
                     files, suggestion, new_image_filenames
                 )
 
+            self._copy_images_from_target_exploration_to_translation(suggestion)
+
         self.render_json(self.values)
 
+    def _upload_new_images_in_translation(
+        self,
+        files: Dict[str, str],
+        suggestion: suggestion_registry.SuggestionTranslateContent,
+        filenames: List[str]
+    ) -> None:
+        """Saves new images introduced in translation suggestion to storage.
+
+        Args:
+            files: dict. Files containing a mapping of image
+                filename to image blob.
+            suggestion: SuggestionTranslateContent. The translation suggestion
+                for which images are being uploaded.
+            filenames: list(str). The image filenames.
+        """
+        suggestion_image_context = suggestion.image_context
+        # TODO(#10513): Find a way to save the images before the suggestion is
+        # created.
+        for filename in filenames:
+            image = files[filename]
+            decoded_image = base64.decodebytes(image.encode('utf-8'))
+            file_format = (
+                image_validation_services.validate_image_and_filename(
+                    decoded_image, filename))
+            image_is_compressible = (
+                file_format in feconf.COMPRESSIBLE_IMAGE_FORMATS)
+            fs_services.save_original_and_compressed_versions_of_image(
+                filename, suggestion_image_context, suggestion.target_id,
+                decoded_image, 'image', image_is_compressible)
+
+    def _copy_images_from_target_exploration_to_translation(
+        self,
+        suggestion: suggestion_registry.SuggestionTranslateContent
+    ) -> None:
+        """Creates copies of images from the suggestion's target exploration
+        for the translation suggestion to use.
+
+        Args:
+            suggestion: SuggestionTranslateContent. The translation suggestion
+                to copy its target exploration's images to.
+
+        Raises:
+            Exception. An image in the target exploration's content is not a
+                saved asset belonging to the target exploration.
+        """
+        target_image_filenames = (
+            html_cleaner.get_image_filenames_from_html_strings(
+                suggestion.get_target_entity_html_strings()
+            )
+        )
+        try:
+            fs_services.copy_images(
+                suggestion.target_type, suggestion.target_id,
+                suggestion.image_context, suggestion.target_id,
+                target_image_filenames
+            )
+        except ValueError as error:
+            raise Exception(
+                'There are images in the target exploration that are '
```

Comment:
This is somewhat unclear and I wouldn't know what to do if I saw this error. Could you please explain it more simply?

Reply:
`An image in the submitted translation's original content named "{filename}" cannot be found. Please save it to the backend file system at /exploration/{exp_id}/assets/image/ before submitting this translation again.`

For example, if an image with filename `img.png` was part of submitted translation's target exploration with id `s5Fr3Zd`, then the error message will read: `An image in the submitted translation's original content named "img.png" cannot be found. Please save it to the backend file system at /exploration/s5Fr3Zd/assets/image/ before submitting this translation again.`

Also, I think I should move the copy operation of old images before the save operation of new images. That way if the images aren't copied successfully, then we don't save the new images either.

@seanlip What do you think?

Reply:
Yup, that's much clearer, thanks!

I don't have a strong opinion on which order we should do the copy/save in -- I'm not sure why we'd want to test copying first in particular. One suggestion to keep things clean might be to fetch all the necessary images first (from both the original content and the new suggestion), and then save all of them to the correct paths in one for loop.

Reply:
@seanlip I restructured the error message.